File: client-old/src/types/schema/operation/composite.ts
import { smoFromServer } from '.';
import type { SchemaCategory } from '../SchemaCategory';
import { type SMO, type SMOFromServer, SMOType } from './smo';

export type CompositeFromServer = SMOFromServer<SMOType.Composite> & {
    name: string;
    children: SMOFromServer[];
};

export class Composite implements SMO<SMOType.Composite> {
    readonly type = SMOType.Composite;

    private constructor(
        readonly name: string,
        readonly children: SMO[],
    ) {}

    static fromServer(input: CompositeFromServer): Composite {
        return new Composite(
            input.name,
            input.children.map(smoFromServer),
        );
    }

    static create(name: string, children: SMO[]): Composite {
        return new Composite(
            name,
            children,
        );
    }

    toServer(): CompositeFromServer {
        return {
            type: SMOType.Composite,
            name: this.name,
            children: this.children.map(smo => smo.toServer()),
        };
    }

    up(category: SchemaCategory): void {
        this.children.forEach(child => child.up(category));
    }

    down(category: SchemaCategory): void {
        [ ...this.children ].reverse().forEach(child => child.down(category));
    }
}

File: library/evolution/src/main/java/cz/matfyz/evolution/category/Composite.java
package cz.matfyz.evolution.category;

import cz.matfyz.core.metadata.MetadataCategory;
import cz.matfyz.core.schema.SchemaCategory;

import java.util.List;

public record Composite(
    String name,
    List<SMO> children
) implements SMO {

    @Override public <T> T accept(SchemaEvolutionVisitor<T> visitor) {
        return visitor.visit(this);
    }

    @Override public void up(SchemaCategory schema, MetadataCategory metadata) {
        for (final var child : children)
            child.up(schema, metadata);
    }

    @Override public void down(SchemaCategory schema, MetadataCategory metadata) {
        for (final var child : children.reversed())
            child.down(schema, metadata);
    }

}
